### PR TITLE
Fix .profile.d $PATH destruction and stderr logging

### PR DESCRIPTION
**Pull request by dzuelke**
The export statement for `$PATH` in the `.profile.d` script writes the current string literal to the script, rather than a dynamic `$PATH` reference.

This results in all `$PATH` values set at startup in scripts that are sourced before `blackfire-agent.sh` to be wiped away and replaced with the build-time value of `$PATH`.

The agent so far has also been logging to a file, making the logs invisible to `heroku logs`. We hard-code the env var to `stderr` and pass it in explicitly.

Side question: what's the order of precedence for config? Can an env var always override any value, whether set by config file or by command line arg? I'd expect the order, in ascending priority, to be 1) config, 2) env var, 3) cmdline arg, is that correct? (I can update the PR to do this more cleanly once I know the answer).

### CODE REVIEW

**Review by lolautruche — APPROVED** [review]

**Comment by lolautruche** [comment]
Thank you @dzuelke !

> Side question: what's the order of precedence for config? Can an env var always override any value, whether set by config file or by command line arg? I'd expect the order, in ascending priority, to be 1) config, 2) env var, 3) cmdline arg, is that correct? (I can update the PR to do this more cleanly once I know the answer).

The order of precedence is indeed what you describe.

If everything if fine for you, I can merge this in :+1:

**Comment by dzuelke** [comment]
I realized my question was ambiguous, since priority "1" commonly means highest (as in "most important"), but I meant "in ascending order of evaluation".

Sorry for the confusion, just trying to nail it down.

An env var can override a config file setting, but not a command line arg, is that right? And a command line arg overrides both?

**Comment by lolautruche** [comment]
The precedence order is, from highest to lowest:
1. CLI args
2. Environment variables
3. Config file

**Comment by dzuelke** [comment]
Perfect. Thanks for clarifying. Let me see if I need to adjust anything in the PR then...

**Comment by dzuelke** [comment]
Yeah I guess this can be merged as-is.

Longer term, it might be cool if the agent could be started without a config file at all, given how all settings are in env vars anyway.

The docs say that the PHP probe will take its config from the CLI (https://blackfire.io/docs/php/configuration#configuring-the-probe-via-the-blackfire-cli-configuration-file):

> When environment variables are not set, the probe reads its default configuration from the configuration of the Blackfire CLI. The location of this file depends on the operating system you use.

Does that refer literally to the CLI agent file, or does it "read" the values through some protocol?

The reason why I'm asking all these config related questions is that I want to avoid situations where a) users can `heroku config:set BLACKFIRE_LOG_FILE` or similar and break things; b) a redeploy is necessary when changing e.g. `BLACKFIRE_LOG_LEVEL` because it's read from a file (written at deploy time) in one and an env var in a different location.

Right now it all works fine I believe, but it might be better to explicitly structure the code with all this in mind.

**Comment by lolautruche** [comment]
> Longer term, it might be cool if the agent could be started without a config file at all, given how all settings are in env vars anyway.

That would indeed be cool. Would it be possible @iamluc ?

> Does that refer literally to the CLI agent file, or does it "read" the values through some protocol?

The documentation you're referring to is actually about the probe, not the agent :wink:. The SDK reads the CLI (client) config file (usually located at `~/.blackfire.ini`) when needed.
As for the agent, the thing is that it must be restarted whenever one changes its configuration, that's why we currently don't have a choice...

I agree with you, changing these shouldn't imply a redeploy...
I know that in other PaaS, a redeploy is implicitly done whenever an env var is added or changed. Isn't it the case with Heroku as well?

**Comment by lolautruche** [comment]
@iamluc just told me that it is actually possible to get rid of the config file by using `--config=/dev/null`.
However, the agent would always need to be restarted whenever we change the config.

**Comment by iamluc** [comment]
> @iamluc just told me that it is actually possible to get rid of the config file by using `--config=/dev/null`.

Or with an env var like at https://github.com/blackfireio/docker/blob/master/Dockerfile#L3

**Comment by dzuelke** [comment]
> As for the agent, the thing is that it must be restarted whenever one changes its configuration, that's why we currently don't have a choice...

If you `heroku config:set`, or `heroku ps:restart`, or `heroku ps:scale`, or deploy a code change, then all dynos (think of them just like Docker containers) restart, so the agent obviously gets "restarted" as well (since it gets executed on startup).

> I agree with you, changing these shouldn't imply a redeploy...
> I know that in other PaaS, a redeploy is implicitly done whenever an env var is added or changed. Isn't it the case with Heroku as well?

This is the case! When you change config, the latest build (from a `git push heroku master` or similar) gets re-started with the updated config vars.

We call this a "release" - any time you make a code change, you get a new release, but also any time you make a config change.

The important distinction is that a config change does not perform a rebuild.

Basically: `latest_built_code + current_config = release`.

Think of it as a `docker build` (that's when the buildpack runs) and a `docker run --env-file=…` (that's running dynos for web, workers, etc).

Does that make sense?

**Comment by dzuelke** [comment]
> @iamluc just told me that it is actually possible to get rid of the config file by using `--config=/dev/null`.

Ah, is that new in v2? I think I tried that with agent v1 and it complained that there was no `[blackfire]` section.

> However, the agent would always need to be restarted whenever we change the config.

Yeah that's not a problem, the buildpack just makes an image (it's a tarball, but conceptually the same as a built Docker image), so any config changes of course mean a re-start of the whole container.

**Comment by lolautruche** [comment]
Ah, nice! So we might manage to avoid a full rebuild :tada: !

Thank you @dzuelke for the clarification :slightly_smiling_face: .
Would you like to update your PR or shall we do the changes and test from another one?

**Comment by dzuelke** [comment]
> > Does that refer literally to the CLI agent file, or does it "read" the values through some protocol?
> 
> The documentation you're referring to is actually about the probe, not the agent 😉. The SDK reads the CLI (client) config file (usually located at `~/.blackfire.ini`) when needed.

Yeah, the probe was what I was referring to. My concern basically is that if the probe INI does nor define a value for something, but the agent config does - what happens if the agent config was overridden with a CLI argument?

Example:
- agent INI has `log-file=/foo/bar`
- there is no `$BLACKFIRE_LOG_FILE` set
- agent was started with `--log-file=stderr`

Does the probe now use `/foo/bar` or `stderr`?

**Comment by dzuelke** [comment]
> Ah, nice! So we might manage to avoid a full rebuild 🎉 !
> 
> Thank you @dzuelke for the clarification 🙂 .
> Would you like to update your PR or shall we do the changes and test from another one?

I think it's fine for now. I can probably make another PR that doesn't even write a config file, but right now, it looks like all of the values that are written to the file are  (maybe by coincidence? see ) also `export`ed as environment variables (or already in the environment from `heroku config`), and because those take precedence, no redeploy is necessary when making changes.

Important note I forgot to mention: a `heroku config` triggers a new release, and that means the buildpack doesn't even run, because no build process runs!

That's why it's so important not to hard-code env vars at build time, but I think it's all good because environment variables override config file settings.

And `$BLACKFIRE_LOG_FILE` and `$BLACKFIRE_AGENT_SOCKET` are hard-coded on startup, which means customers cannot accidentally break things there.

**Comment by lolautruche** [comment]
The probe and the agent are configured completely separately. They do share some environment variables though (e.g. `$BLACKFIRE_LOG_FILE`, `$BLACKFIRE_LOG_LEVEL`.

> Example:
>
>    agent INI has log-file=/foo/bar
>    there is no $BLACKFIRE_LOG_FILE set
>    agent was started with --log-file=stderr
>
> Does the probe now use /foo/bar or stderr?

The probe will use the config from `php.ini` (or any ini file loaded by PHP)

**Comment by lolautruche** [comment]
> it looks like all of the values that are written to the file are (maybe by coincidence? see ) also exported as environment variables (or already in the environment from heroku config), and because those take precedence, no redeploy is necessary when making changes.

This is correct for the env vars which are exported (IIRC, this is why I exported them in this way).
However, this is not the case for the credentials, yet. 

Let's merge this!
Thank you so much @dzuelke

**Comment by dzuelke** [comment]
Hmh, but the docs say it uses the agent config, @lolautruche:

> When environment variables are not set, the probe reads its default configuration from the configuration of the Blackfire CLI.

**Comment by dzuelke** [comment]
> This is correct for the env vars which are exported (IIRC, this is why I exported them in this way).
> However, this is not the case for the credentials, yet.

It's also the case for the credentials! They are `export`ed into the environment when the container starts, so they'll also override the values written to the config file.

The env vars in `.profile.d/` all set reasonable default values (log level etc).

I guess the one corner case then is if someone runs `heroku config:remove BLACKFIRE_SERVER_ID`, then it would fall back to whatever the value was when last deployed, meaning you couldn't "disable" Blackfire quickly without a redeploy.

**Comment by lolautruche** [comment]
It is confusing, I agree, but Blackfire CLI is not necessarily the agent (agent is a sub-command).
Let me fix the doc :wink:

**Comment by lolautruche** [comment]
> I guess the one corner case then is if someone runs heroku config:remove BLACKFIRE_SERVER_ID, then it would fall back to whatever the value was when last deployed, meaning you couldn't "disable" Blackfire quickly without a redeploy.

In that case, the build would fail I think.

**Comment by dzuelke** [comment]
> > I guess the one corner case then is if someone runs heroku config:remove BLACKFIRE_SERVER_ID, then it would fall back to whatever the value was when last deployed, meaning you couldn't "disable" Blackfire quickly without a redeploy.
> 
> In that case, the build would fail I think.

There is no build that runs in that case!

Example:

```ShellSession
$ mkdir heroku-blackfire-env-vars-test
$ cd heroku-blackfire-env-vars-test
$ git init
$ touch index.php
$ git add .
$ git commit -m "import"
$ heroku create
$ heroku buildpacks:set https://github.com/blackfireio/integration-heroku
$ heroku buildpacks:add heroku/php
$ heroku config:set BLACKFIRE_SERVER_ID=foo BLACKFIRE_SERVER_TOKEN=bar
$ git push heroku master # this builds a new so-called "slug"
$ heroku run env # blackfire agent now uses those "foo"/"bar" env vars, and they're also written to the config file
$ heroku config:set BLACKFIRE_SERVER_ID=one BLACKFIRE_SERVER_TOKEN=two # this restarts the app with the slug from last git push but updated env vars
$ heroku run env # blackfire agent now uses those "one"/"two" env vars instead of the config file values
$ heroku config:remove BLACKFIRE_SERVER_ID BLACKFIRE_SERVER_TOKEN # this restarts the app with the slug from last git push but these two env vars removed
$ heroku run env # the env vars are gone but blackfire agent uses "foo" and "bar" again since those were the values during last deploy and they are still in agent config
```

Does that make sense?

**Comment by lolautruche** [comment]
Yes it does, thanks @dzuelke

**Comment by dzuelke** [comment]
Alright I'll see if I have time later to make a PR that moves to `/dev/null` and env-vars-only for config. It would mean having to duplicate a bunch of the "is `$BLACKFIRE_SERVER_ID` there?" etc checks from build-time to run-time, although maybe the error messages in `heroku logs` from the agent on startup will be enough.

**Comment by dzuelke** [comment]
Once significant change would be that `agent:config` and maybe also `client:config` don't have to be called anymore, meaning the credentials aren't validated.

We can call `agent:start --test` to validate the config, but what about the client?
